### DISCUSSION TITLE
Current module view not available in newest interface

### DISCUSSION THREAD

**Discussion by dlong500** [codesandbox/codesandbox-client#8350] [General]
I noticed within the last couple weeks or so that some open source projects that have set up examples via a CodeSandbox project and made use of the "current module view" feature so that the preview window will render the contents of the active editor (rather than just the project entry page) have stopped working and there is apparently no longer a toggle button for `Project view` vs `Current module view`.

I was able to find a workaround by disabling the `Sandbox beta editor` setting, but even that required opening a new window and going to the original URL for the projects in question as it appears that any CodeSandbox URL starting with `https://codesandbox.io/p/` use the new interface no matter what the `beta editor` setting is.  Also, if you aren't signed into CodeSandbox then it appears ALL URLs now forward to the new `/p/` type URL structure.

Is the "current module view" going to be added to the latest interface?  This is an important feature for open source projects as it avoids having to set up a separate CodeSandbox project for each demo.

See https://github.com/downshift-js/downshift/issues/1570 for context with actual real world usage.

**Comment by JamesACS** [comment]
Hello @dlong500 With our new editor, you can specify the default file for the editor to open on in an embed using the src attribute -  you can see this in the example in our [docs](https://codesandbox.io/docs/learn/sandboxes/embedding). Does this solve the issue for you?

**Comment by dlong500** [comment]
@JamesACS Unfortunately that doesn't allow for a hosting a project with a bunch of examples so that you can easily switch between files.  You'd have to create a separate embed for each demo which entails a lot of boilerplate and time to get set up (which kind of defeats the purpose of sites like CodeSandbox),

Also, ideally you would be able to use current module view in a sandbox project even without embedding the sandbox at all so the full interface is available (as you could in the previous interface).

Again, see the downshift project as a real world example.  The sandbox for their example demos is:
https://codesandbox.io/s/github/kentcdodds/downshift-examples

Under the old interface (using current module view) you could simply open one of the example files in the editor, refresh the preview, and the file that is active in the editor would render.  Switching to a different example was as easy as clicking on a different file.

**Reply by CompuIves** [reply]
Yeah, it's a good point. The "Current Module View" was never considered very stable, I added it to the first version of CodeSandbox to allow rendering components individually (and for development on single components) but it broke HMR pretty often. Because of the flakiness of the feature, we didn't enable it in the latest editor.

However, the functionality is still there! So maybe the compromise is that we don't show the button anymore, but we do enable the functionality if the query parameter is there in the URL. That should not be too challenging to add. Let me check...

**Reply by CompuIves** [reply]
Hmm, I see, it's not just a URL parameter but also a UI. Thinking... I remember that Kent is the only one who has used this functionality in the past (from what I know), I would ideally not want to clutter the UI too much for such a small case.

Hmm, tricky, since the example itself also mentions to press a button in a specific place in the UI. I think we could do two things for enabling this functionality:

1. Enable current module view from a URL parameter (this is the easiest option, less state management)
2. Enable current module view from command palette.

I think the first option would be the easiest, then we change the URL to the sandbox to add `?currentModuleView`, which will automatically enable that mode.

**Reply by dlong500** [reply]
So with the potential URL parameter would that be persistent when clicking on a different file in a sandbox?  As long as current module view persisted when browsing around a sandbox so the newly selected file would render in the preview window then I think there would be no need for a UI switch.

Thanks for looking into this!

**Reply by krvajal** [reply]
this is broken https://codesandbox.io/p/sandbox/github/kentcdodds/downshift-examples/tree/master/

**Reply by dlong500** [reply]
@krvajal Yes, that is the reason for the my original post.  A potential fix for the new interface was mentioned but I don't think there have been any updates yet.  @CompuIves can you provide an update on where things stand?

**Reply by CompuIves** [reply]
Right! We had to make an update to the Sandpack API to make this work, now it's a matter of connecting the URL params to use that API. I'm going to take a look tomorrow!

**Reply by dlong500** [reply]
@CompuIves Sorry to pester, but are there any new updates on this?  I've noticed several more people having difficulty trying to view the downshift examples and it seems the workaround to use the old editor may be broken.  If the option to use the old editor goes away then any project depending on currentModuleView for an example repository will be permanently broken.

**Reply by CompuIves** [reply]
Absolutely no pester! I forgot to follow up on this and will do it right _now_! 😄

**Reply by CompuIves** [reply]
Okay! PR has been opened. I expect to get a review tomorrow so the fix will be live then. It's important to note that you should add `moduleview=1` as a query parameter to the URL to enable this functionality. I opted to not put it in the UI as its use case is super small, and it would clutter the UI more. I hope that's fine!

**Reply by dlong500** [reply]
That's great!  The only question I would have is will the query parameter persist when clicking on different modules within the project?  As long as the parameter persists then I don't see a problem without a UI switch.

**Reply by dlong500** [reply]
@CompuIves The `moduleview=1` query parameter works well, so thanks for that!  One little issue, however, is that unless an external link already contains the parameter it would seem the only way to add it is with the user manually adding it.  There appears to be no way to set the parameter with a link on an index page inside the codesandbox project, as the embedded preview window is sandboxed without the `allow-top-navigation` flag set.  Is there a way to generate a link within a codesandbox that would add that parameter programatically to the top-level window?  Another possibility might be if we could add an option for a codesandbox project to default to the module view.

Sorry to keep bugging you but I feel like we're 95% there.  It's usable now but could cause a bit of friction with users needing to manually add the query parameter after reaching the codesandbox page.

**Reply by coreybruyere** [reply]
Running into the same issue. Not sure I understand where to use ?moduleview=1 param for viewing individual modules on downshift examples page. Care to provide some more info? Thanks

**Reply by dlong500** [reply]
If you are on the downshift codesandbox page using the newest interface the URL structure for a GitHub repository-based sandbox likely already has a `file` query parameter like so:
https://codesandbox.io/p/sandbox/github/kentcdodds/downshift-examples/tree/master/?file=%2Fsrc%2Findex.js
You can append another query parameter with an `&`, so it would become:
https://codesandbox.io/p/sandbox/github/kentcdodds/downshift-examples/tree/master/?file=%2Fsrc%2Findex.js&moduleview=1

At that point you can now navigate to the various example modules and the module view will persist (thus running the code in the preview pane).

**Reply by luancardosolc** [reply]
Thanks for tip @dlong500, it helped me so much!

**Comment by CompuIves** [comment]
Interesting, do you have the ability to change the sandbox contents? I think I can also enable it if a certain config is inside the sandbox.

**Reply by dlong500** [reply]
I'm not a maintainer.  I can submit a PR to the downshift-examples GitHub repo, but it doesn't look like there are codesandbox settings in the actual examples repo.  I would hope the current downshift maintainer could update the sandbox settings if necessary.
